src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs

```diff
@@ -191,22 +189,32 @@ private static FileStream CreateSharedBackingObject(Interop.Sys.MemoryMappedProt
             if ((protections & Interop.Sys.MemoryMappedProtections.PROT_EXEC) != 0)
                 perms |= Interop.Sys.Permissions.S_IXUSR;
 
-            // Create the shared memory object.
-            SafeFileHandle fd = Interop.Sys.ShmOpen(mapName, flags, (int)perms);
-            if (fd.IsInvalid)
+            string mapName;
+            SafeFileHandle fd;
+
+            do
             {
-                Interop.ErrorInfo errorInfo = Interop.Sys.GetLastErrorInfo();
-                if (errorInfo.Error == Interop.Error.ENOTSUP)
+                mapName = GenerateMapName();
+                fd = Interop.Sys.ShmOpen(mapName, flags, (int)perms); // Create the shared memory object.
+
+                if (fd.IsInvalid)
                 {
-                    // If ShmOpen is not supported, fall back to file backing object.
-                    // Note that the System.Native shim will force this failure on platforms where
-                    // the result of native shm_open does not work well with our subsequent call
-                    // to mmap.
-                    return null;
-                }
+                    Interop.ErrorInfo errorInfo = Interop.Sys.GetLastErrorInfo();
+                    fd.Dispose();
```

Comment:
FWIW, it's a little weird to Dispose of the SafeHandle and then continue to use it, but technically in this case it's ok to do so, given how it's being used.  It's just unusual to see and thus a little disconcerting.

Reply:
Maybe a comment here would help clarify the unusual usage?